Libbeat's HTTP Server can now listen to unix socket.

ph opened the pull request:
Allow to use a socket file using the `unix:///tmp/hello.sock` syntax
to define an HTTP server that will listen HTTP request and add support for windows named pipe using the `npipe:///hello` syntax.

Tasks:

- Implements the host parsing.
- Refactor the API handler to be reusable and testable.
- Refactor the API Code so we can start and stop a server gracefully,
  this is require to clean up the socket after usage.
- Allow to mock monitoring.

**NOTES**: I need to verify the SecurityDescriptor permission for the named pipe.
Fixes: #13577

cachedout commented on filebeat/filebeat.reference.yml:
```diff
@@ -2034,7 +2034,8 @@ logging.files:
 
-# The HTTP endpoint will bind to this hostname or IP address. It is recommended to use only localhost.
+# The HTTP endpoint will bind to this hostname, IP address or unix socket. It is recommended
+# to use only localhost.
```
As far as the recommendation goes, what if it just said `If binding to an IP address it is recommended to use only localhost.`?

ph replied:
@cachedout I think the wording is fine, but I presume @andrewkroh question was more should we default to npipe or Unix socket instead of TCP for the HTTP endpoint.

ycombinator replied:
My vote would be to default to the unix socket/npipe (starting 8.0.0 with deprecation warnings in 7.x, if possible) since its more secure (can only accept connections from local machine) and slightly faster than TCP. If users need the HTTP API to be callable from outside the local machine, they can take the deliberate step of configuring that.

ph replied:
@ycombinator @cachedout What about the "beta" warning when starting the API?

ycombinator replied:
You mean this line?

```
2019-09-13T08:52:45.092-0700    WARN    [cfgwarn]       api/server.go:34        EXPERIMENTAL: Metrics endpoint is enabled.
```

If so, I vote for it to be taken out. As you said in another comment, this API has been around for a while now.

ph replied:
I've removed the beta label but I didn't add a deprecation or changed the defaults.

michalpristas commented on libbeat/api/npipe/pipe.go:
```diff
@@ -0,0 +1,25 @@
+
+// IsNPipe returns true if the string has a npipe scheme.
+func IsNPipe(s string) bool {
+	return strings.HasPrefix(s, "npipe:///")
```
what's your opinion on supporting also `HasPrefix("//./pipe/")`?

ph replied:
I am ok with supporting both, to be honest, I was not sure what would be the best way to represent them in a configuration, but supporting npipe + the real representation \\.\pipe\hello is ok.

ycombinator commented:
I built Filebeat with this PR and then added the following lines to my `filebeat.yml`:

```
http.enabled: true
http.host: unix:///tmp/foo.sock
```

When I run Filebeat with `filebeat -e`, I get these messages in the log:

```
2019-09-13T08:52:45.092-0700    INFO    api/server.go:38        Starting stats endpoint
2019-09-13T08:52:45.093-0700    INFO    api/server.go:49        Metrics endpoint listening on: unix:///tmp/foo.sock:5066
2019-09-13T08:52:45.093-0700    INFO    api/server.go:51        finished starting stats endpoint: listen tcp: address unix:///tmp/foo.sock:5066: too many colons in address
```

Two questions:

1. Am I specifying the Unix socket incorrectly in my configuration?
2. Shouldn't the "too many colons in address" log message be of type ERROR instead of INFO?

ph commented:
>    Am I specifying the Unix socket incorrectly in my configuration?
>port should be ignored, I have a similar test config and it just pass. I will check if I can add a system test.

> Shouldn't the "too many colons in address" log message be of type ERROR instead of INFO?
Good point, I think the statement was before the refactoring.

ph commented on libbeat/api/server.go:
```diff
@@ -19,86 +19,93 @@ package api
+	}
 
-	data := monitoring.CollectStructSnapshot(monitoring.GetNamespace("stats").GetRegistry(), monitoring.Full, false)
+	l, err := net.Listen(network, path)
```
Good catch, I think we should clean up for the unix socket, especially if there is a crash and we try to restart it we could be in an infinite look.

Concerning the permission the defaultis the following:

srwxr-xr-x 1 ph   users     0 Sep 13 17:14 foo.sock

We might want to remove the x from all and keep only user/group.

ph commented on libbeat/api/server.go:
```diff
@@ -51,16 +56,16 @@ func New(mux *http.ServeMux, config *common.Config) (*Server, error) {
 	}
 
-	return &Server{mux: mux, l: l, config: cfg}, nil
+	return &Server{mux: mux, l: l, config: cfg, log: log.Named("API")}, nil
```
I've downcased the api.

>  This will create a new logger and append .API to the existing logger's name (if it had a name).

Do you find the append problematic?

Not sure adding the listener or endpoint is useful, we only have a single responder?

ph commented:
@ycombinator lets pair on monday, I was not able to reproduce your error, are you using the latest commits?

ph commented:
I need to add support for the helper.NewHTTP to connect to both unix socket and named pipe.

ycombinator commented:
> @ycombinator lets pair on monday, I was not able to reproduce your error, are you using the latest commits?

I will try again and if I can still reproduce the same error, lets pair on Monday.

ycombinator commented on filebeat/filebeat.yml:
```diff
@@ -215,3 +215,5 @@ processors:
 # This allows to enable 6.7 migration aliases
 #migration.6_to_7.enabled: true
+http.enabled: true
+http.host: unix:///tmp/foo.sock:111111
```
Did you mean to commit this change?

ph replied:
I did not.

ycombinator commented:
> I will try again and if I can still reproduce the same error, lets pair on Monday.

I checked out this PR again and now things are working as expected.

ycombinator commented on libbeat/api/make_listener_windows.go:
```diff
@@ -0,0 +1,54 @@
+
+	if network == "unix" {
+		return nil, fmt.Errorf(
+			"cannot use %s as the host, unix socket are only supported on Linux and macOS platform",
```
Nit typos: socket => sockets, platform => platforms

ph commented:
@ycombinator do you mind if I split server and client changes in another PR?  That will make it smaller to review.

ycombinator commented on libbeat/api/server.go:
```diff
@@ -19,86 +19,84 @@ package api
-	} else {
-		fmt.Fprintf(w, data.String())
-	}
+func isWindows() bool {
```
Should this be moved into `server_test.go`? Seems like it's the only place that really needs it?

ycombinator commented:
> do you mind if I split server and client changes in another PR? That will make it smaller to review.

Not at all, even better! Thanks @ph.

ph commented:
I believe I have addressed the concerns of @ycombinator and @andrewkroh.

Concerning the changing the default and adding a deprecation warning, I would prefer we have this out of the door with a few usage from users. So we can do it in 7.5

ycombinator commented:
> Concerning the changing the default and adding a deprecation warning, I would prefer we have this out of the door with a few usage from users. So we can do it in 7.6

++ Also, we might want to separately discuss whether this needs to go into 8.0.0 only since we're changing the default behavior. It's not a breaking change per se but it may catch users by surprise. Anyway, let's discuss that in a separate issue/PR from this one.

ycombinator commented on libbeat/api/npipe/listener_windows.go:
```diff
@@ -0,0 +1,79 @@
+		return "", errors.Wrap(err, "failed to retrieve the current user")
+	}
+
+	// TODO(ph): I have to review this before final review.
```
@ph Did you get a chance to review this? 😄

cachedout commented on auditbeat/auditbeat.reference.yml:
```diff
@@ -1333,7 +1333,8 @@ logging.files:
 #http.enabled: false
 
-# The HTTP endpoint will bind to this hostname or IP address. It is recommended to use only localhost.
+# The HTTP endpoint will bind to this hostname, IP address, unix socket or named pipe. It is recommended
```
This doesn't quite make sense any longer. There's no concept of "localhost" for unix sockets. What if it read something like, `The HTTP endpoint will bind to this hostname, IP address, unix socket or named pipe. When using IP addresses, it is recommended to only use localhost`?

[This change would need to go in a few places in this PR if accepted.]

ph replied:
changed in all the yaml.

cachedout commented on libbeat/api/make_listener_posix.go:
```diff
@@ -0,0 +1,78 @@
+func makeListener(host string, port int) (net.Listener, error) {
+	if npipe.IsNPipe(host) {
+		return nil, fmt.Errorf(
+			"cannot use %s as the host, named pipe are only supported on Windows",
```
s/named pipe/named pipes/

cachedout commented on libbeat/api/npipe/listerner_posix.go:
```diff
@@ -0,0 +1,33 @@
+// DialContext create a Dial to be use with an http.Client to connect to a pipe.
+func DialContext(npipe string) func(context.Context, string, string) (net.Conn, error) {
+	return func(ctx context.Context, _, _ string) (net.Conn, error) {
+		return nil, errors.New("named pipe doesn't work on linux")
```
Doesn't linux support named pipes using `mkfifo`?

ph replied:
They are supported through `syscall.mkfifo` and I think we could support them but that would require a bit of close to the metal implementation and should be similar to what go-winio is doing in https://github.com/microsoft/go-winio/blob/84b4ab48a50763fe7b3abcef38e5205c12027fac/pipe.go#L185-L203. Since they aren't working out of the box, and I don't know all the gotcha for the implementation I consider them not required for the first version of this implementation.

urso replied:
I'm okay with postponing mkfifo.

Consider error wording: "names pipes are only supported on Windows".

cachedout replied:
Thanks for the explanation, @ph. I'll mark this as resolved.

michalpristas commented on libbeat/api/npipe/listener_windows.go:
```diff
@@ -0,0 +1,77 @@
+		return "", errors.Wrap(err, "failed to retrieve the current user")
+	}
+
+	descriptor := "D:P(A;;GA;;;" + u.Uid + ")"
```
this gives user read/write access to the pipe.
maybe we can repalce Generic REad write to: FAFR
also we might think about passing a user id from config/env. this will open a pipe only for a user beat is running as. 
consider an agent starting a beat as UID1 and then a special metricbeat to monitor this beat as UID2. this setting will force UID1==UID2.

urso replied:
Would it make sense to make the uid configurable?

We also need some comment (with link to windows docs), why the descriptor is chosen as is.

ph replied:
Yeah, I believe it make sense to allow it to be configurable and maybe a fallback to the current using when not configured would be the way to go. I am going to make the change @michalpristas / @elastic/solutions-docs 

Thanks for looking at it, the documentation for the security descriptor is not the best.

urso commented on libbeat/api/make_listener_windows.go:
```diff
@@ -0,0 +1,54 @@
+
+	if network == "unix" {
+		return nil, fmt.Errorf(
+			"cannot use %s as the host, unix sockets are only supported on Linux and macOS platforms",
```
From build/usage of APIs it looks like unix sockets will be supported by other platforma as well.
Let's change the error message to:

"unix sockets are not supported on Windows, use npipe instead".

ph commented:
The failures on TravisCI seems to be related to https://github.com/elastic/beats/pull/13987
Beats-ci is mostly green, please do a final review so we can merge this. Thanks.

ycombinator commented on auditbeat/auditbeat.reference.yml:
```diff
@@ -1333,12 +1333,21 @@ logging.files:
 #http.port: 5066
 
+# Define which user should be owning the named pipe.
+#http.user:
```
Would it make sense to place this setting and the one right below it under a `http.named_pipe` (or similar) namespace?

ph replied:
Make sense, I think it will make it more obvious for the user.

ycombinator commented on libbeat/api/make_listener_posix.go:
```diff
@@ -0,0 +1,83 @@
+
+	// Ensure file mode
+	if network == "unix" {
+		if err := os.Chmod(path, socketFileMode); err != nil {
```
Perhaps this check could be done even before we start listening?  It could be made part of the previous `if network == "unix"`block?

ph replied:
We can't the socket create the file and we change the mode on it. its not ideal, I've tried using fchmod instead with the original fd and everytime I did that it just block/freeze. I went down a rabbit hole.

ycombinator replied:
Ah duh, my stupidity! The socket file won't exist until `net.Listen` succeeds. 🤦‍♂

ycombinator commented on libbeat/api/server.go:
```diff
@@ -19,86 +19,77 @@ package api
+	case "http":
+		return "tcp", url.Host, nil
+	case "unix":
+		return url.Scheme, url.Path, nil
```
Not sure if we should care about this but just want to check.

Using `-E http.host=unix:///tmp/mb.sock` and `curl -v --unix-socket /tmp/mb.sock 'http://unix:5434/'` works just fine. Should we return an error if the scheme is `unix` and a port is set as well?

ycombinator commented on libbeat/api/make_listener_posix.go:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	if network == "unix" {
+		if _, err := os.Stat(path); !os.IsNotExist(err) {
```
The double negative if predicate, `!os.IsNotExist(err)` is a bit hard to parse. Any reason we couldn't do `os.IsExist(err)`?

ycombinator replied:
Just chatted with @ph off-PR about this change so wanted to summarize our discussion here. Apparently, if the double-negative check here is replaced with `os.IsExist(err)`, things don't work as expected:

```
$ touch /tmp/foo
$ ./filebeat -c filebeat.dev.yml -E http.host=unix:///tmp/foo                                                                                                                                      
Exiting: could not start the HTTP server for the API: listen unix /tmp/foo: bind: address already in use
```

But if we keep the double-negative check, everything works as expected. So we are bringing it back.

ph commented:
@ycombinator updated with your comments.

ruflin commented:
Just stumbled over this, great to see this implemented. I wonder if this could allow us to have the endpoint enabled by default (if not already the case)?

ph commented:
@ruflin I think it could be on by default but maybe it should be included in the 8.0 deprecation or changes issue? wdyt @ycombinator.

ph commented:
Lets  move the discussion in https://github.com/elastic/beats/issues/12918
